File: 90-subsets-ii/subsets-ii.ts
function subsetsWithDup(nums: number[]): number[][] {
    const ans: number[][] = [];
    const ds: number[] = [];

    function findSubsets(ind: number): void {
        ans.push([...ds]);
        for (let i = ind; i < nums.length; i++) {
            if (i !== ind && nums[i] === nums[i - 1]) {
                continue;
            }
            ds.push(nums[i]);
            findSubsets(i + 1);
            ds.pop();
        }
    }

    nums.sort((a, b) => a - b); // Sort to handle duplicates
    findSubsets(0);
    return ans;
}

File: 90-subsets-ii/subsets-ii.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> curList = new ArrayList<>();
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        generate(nums , 0);
        return ans;
    }
    private void generate(int[] nums , int curIndex){
        if(curIndex == nums.length){
            ans.add(new ArrayList(curList));
            return;
        }


        // pick
        curList.add(nums[curIndex]);
        generate(nums , curIndex + 1);
        curList.remove(curList.size() - 1);
        // don't pick
        int tempIndex = curIndex;
        while(tempIndex != nums.length - 1 && nums[tempIndex] == nums[tempIndex + 1]) tempIndex ++;
        generate(nums , tempIndex + 1);
    }
}